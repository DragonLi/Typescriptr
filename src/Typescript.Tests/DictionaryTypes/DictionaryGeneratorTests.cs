using System.Collections.Generic;
using Assent;
using Typescriptr;
using Xunit;

namespace Typescript.Tests.DictionaryTypes
{
    public class DictionaryGeneratorTests
    {
        class TypeWithDictionaryProp
        {
            public Dictionary<string, int> DictProp { get; set; }
        }

        [Fact]
        public void Generator_TypeWithDirectDictionaryProp_ShouldRenderToTypescriptMap()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithDictionaryProp)});

            this.Assent(generated.Types);
        }

        class ComplexType
        {
            public string AProp { get; set; }
        }
        
        class TypeWithComplexDictionaryValue
        {
            public Dictionary<int, ComplexType> DictProp { get; set; }
        }

        [Fact]
        public void Generator_TypeWithComplexDictionaryValueType_ShouldRenderToTypescriptMap()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithComplexDictionaryValue)});

            this.Assent(generated.Types);
        }

        class TypeWithCustomDictionaryProp
        {
            public ICustomDictionary DictProp { get; set; }
        }

        interface ICustomDictionary : IDictionary<string, int>
        {
        }
        
        [Fact]
        public void Generator_TypeWithCustomDictionaryValueType_ShouldRenderToTypescriptMap()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithCustomDictionaryProp)});

            this.Assent(generated.JoinTypesAndEnums());
        }

        enum TestEnum
        {
            FirstValue,
            SecondValue,
            ThirdValue,
        }

        class TypeWithDictionaryAndEnumPropAsValue
        {
            public Dictionary<string, TestEnum> EnumDictionary { get; set; }
        }
        
        [Fact]
        public void Generator_TypeWithEnumValueType_ShouldRenderToTypescriptMap()
        {
            var generator = TypeScriptGenerator.CreateDefault();
            var generated = generator.Generate(new[] {typeof(TypeWithDictionaryAndEnumPropAsValue)});

            this.Assent(generated.JoinTypesAndEnums());
        }
    }
}