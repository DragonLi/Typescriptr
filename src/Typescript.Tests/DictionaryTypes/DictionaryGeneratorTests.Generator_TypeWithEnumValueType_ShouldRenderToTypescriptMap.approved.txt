declare namespace Api {
  interface TypeWithDictionaryAndEnumPropAsValue {
    enumDictionary: { [key: string]: 'FirstValue' | 'SecondValue' | 'ThirdValue' };
  }
}

---
function mkenum<T extends {[index: string]: U}, U extends string>(x: T) { return x; }

const TestEnum = mkenum({
  FirstValue: 'FirstValue',
  SecondValue: 'SecondValue',
  ThirdValue: 'ThirdValue',
});
export type TestEnum = (typeof TestEnum)[keyof typeof TestEnum];

